### PR TITLE
bpo-44986: Fixed bug with date formats like %Y-%m-%d in help messages of argparse

### PR DESCRIPTION
**Pull request by SengerM**
# Fixed bug with date formats like %Y-%m-%d in help messages of argparse

Now this code does not produce any error when invoked with the `-h` option and produces the expected result:
```
import argparse

parser = argparse.ArgumentParser(description = 'Just a test code.')
parser.add_argument(
	'--from',
	metavar = 'when', 
	help = 	'A date in format %Y-%m-%d',
	dest = f'From',
	type = str,
	required = True,
)
args = parser.parse_args()

print(args.From)
```

https://bugs.python.org/issue44986

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

## Recognized GitHub username

We couldn't find a [bugs.python.org](https://bugs.python.org/) (b.p.o) account corresponding to the following GitHub usernames:

@SengerM

This might be simply due to a missing "GitHub Name" entry in one's b.p.o account settings. This is necessary for legal reasons before we can look at this contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.




You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Comment by SengerM** [comment]
I don't know if a human will ever read this, but if so I have already signed the CLA yesterday and in the [check yourself](https://check-python-cla.herokuapp.com/) page still says that I have not signed it. Is it normal that it takes so long?

File: Lib/test/test_argparse.py [hunk 1]
```diff
@@ -3901,7 +3901,7 @@ class TestHelpVariableExpansion(HelpTestCase):
     parser_signature = Sig(prog='PROG')
     argument_signatures = [
         Sig('-x', type=int,
-            help='x %(prog)s %(default)s %(type)s %%'),
+            help='x %(prog)s %(default)s %(type)s %'),
```
Comment (rhettinger) [inline Lib/test/test_argparse.py]: This is a breaking change.

Comment (SengerM) [reply Lib/test/test_argparse.py]: I have reverted this change and added a line to replace the escaped `%%` in the `_expand_help` method so now it should be fully compatible with the old string formatting too.

**Review by MaxwellDupre — APPROVED** [review]
Ran the argparse test module:
Ran 1670 tests in 3.185s
OK
== Tests result: SUCCESS ==

And example in bug ok.

**Comment by furkanonder** [comment]
@SengerM Can you add the unit tests for this case?

**Comment by SengerM** [comment]
Dear @furkanonder , sorry for the long delay. I added the test case. I see there are fails in the tests for macOS, but they don't seem related to my changes.

File: Lib/argparse.py [hunk 2]
```diff
@@ -600,7 +600,11 @@ def _expand_help(self, action):
         if params.get('choices') is not None:
             choices_str = ', '.join([str(c) for c in params['choices']])
             params['choices'] = choices_str
-        return self._get_help_string(action) % params
+        expanded_help = self._get_help_string(action)
+        for param, value in params.items():
+            if f'%({param})s' in expanded_help:
```
Comment (savannahostrowski) [inline Lib/argparse.py:605]: I think this conditional is redundant, right? Doesn't `replace()` handle validating that the string is contained by `expanded_help`?

File: Lib/test/test_argparse.py [hunk 3]
```diff
@@ -3934,6 +3934,39 @@ class TestHelpUsageWithParentheses(HelpTestCase):
         '''
     version = ''
 
+class TestHelpWithPercentageSymbols(HelpTestCase):
+    """Test a help message including % symbols"""
```
Comment (savannahostrowski) [inline Lib/test/test_argparse.py:3958]:
```suggestion
    """Test help output that includes % symbols"""
```

File: Lib/test/test_argparse.py [hunk 4]
```diff
@@ -3934,6 +3934,39 @@ class TestHelpUsageWithParentheses(HelpTestCase):
...
+    parser_signature = Sig(prog='PROG', description='Just a test code.')
```
Comment (savannahostrowski) [inline Lib/test/test_argparse.py:3959]:
```suggestion
    parser_signature = Sig(prog='PROG', description='Just a test program.')
```

File: Lib/test/test_argparse.py [hunk 5]
```diff
@@ -3934,6 +3934,39 @@ class TestHelpUsageWithParentheses(HelpTestCase):
...
+    argument_signatures = [
+        Sig('--somearg', metavar='somearg',
+            help='Now you dont need to escape this: %, and you will not get nonsensical errors!'),
```
Comment (savannahostrowski) [inline Lib/test/test_argparse.py:3962]:
```suggestion
            help='No escape required for %'),
```

File: Lib/test/test_argparse.py [hunk 6]
```diff
@@ -3934,6 +3934,39 @@ class TestHelpUsageWithParentheses(HelpTestCase):
...
+        Sig('--date', metavar='when',  help='A date in format %Y-%m-%d',
```
Comment (savannahostrowski) [inline Lib/test/test_argparse.py:3963]:
```suggestion
        Sig('--date', metavar='when',  help='A date in %Y-%m-%d format',
```

File: Lib/test/test_argparse.py [hunk 7]
```diff
@@ -3934,6 +3934,39 @@ class TestHelpUsageWithParentheses(HelpTestCase):
...
+            dest=f'date', type=str, required=True),
+        Sig('bar', nargs='?', type=int, default=42,
+            help='the bar to %(prog)s (default: %(default)s)'),
```
Comment (savannahostrowski) [inline Lib/test/test_argparse.py:3966]:
```suggestion
            help='The bar to %(prog)s (default: %(default)s)'),
```

**Review by savannahostrowski — CHANGES_REQUESTED** [review]
I know this PR has been sitting for a little while, but I pulled it down and rebased it with the main, and it still resolves the bug. Just a couple of comments but overall, this LGTM.

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]
It's not that simple.

What if the replacement value contains `%%` or `%(prog)s`? Or only the part of it, and the other part is contained in the help string? E.g. `help='%(default)s(prog)s'`, where `default='%'`.

What if the help string contains `%(...)r` or `%(...)a` instead of `%(...)s`?

**Comment by serhiy-storchaka** [comment]
`argparse` now checks help string for correctness, so you will be informed about errors earlier.

You can implement a "safe" help formatter as a separate class, but this is not easy task. I am not sure that `argparse` needs this. It is easier to fix help strings in the user code.

**Pull request closed without merging**